### PR TITLE
RendezvousHash#find_nodes

### PR DESCRIPTION
**Pull request by swrobel**
Allow picking `n` nodes with the highest score rather than only the top one.

### CODE REVIEW

**Comment by ewdurbin** [comment]
@swrobel i'd like to better understand the use case for exposing the `find_nodes` method before merging this. clandestined is designed with the goal of determinism, exposing methods for choosing "top two" or "top N" scores isn't well aligned with that, as these choices must be communicated out of band between clients.

**Comment by swrobel** [comment]
My understanding, admittedly based on the [wikipedia article on rendezvous](https://en.wikipedia.org/wiki/Rendezvous_hashing), is that

> Rendezvous or Highest Random Weight (HRW) hashing is an algorithm that allows clients to achieve distributed agreement on a set of k options out of a possible set of n options.

I started my search for an algorithm with [consistent hashing](https://en.wikipedia.org/wiki/Consistent_hashing) but it only chooses 1 option of n, rather than k of n, hence I ended up at rendezvous. I was excited to find your library, but then discovered that it only supported 1 of n and decided to make this enhancement. Feel free to correct me if I'm wrong as I just started looking into consistent hashing algos yesterday.

**Comment by ewdurbin** [comment]
@swrobel I can agree that selecting top `k` results is a valuable by the pure definition of a Rendezvous Hash. The tests pass for the existing interfaces, but I'll need to spend some time making sure we're exposing the "top scores" selection in a good way. Also, I'll need to ensure parity in the other reference implementation; https://github.com/ewdurbin/clandestined-python.

File: lib/clandestined/rendezvous_hash.rb [hunk 1]
```diff
@@ -29,18 +29,17 @@ def remove_node(node)
...
+    def find_nodes(key, num_nodes=1)
+      results = []
       nodes.each do |node|
         score = hash_function.call("#{node}-#{key}")
-        if score > high_score
-          high_score, winner = score, node
-        elsif score == high_score
-          high_score, winner = score, [node.to_s, winner.to_s].max
-        end
+        results << [score, node]
```
Comment (ewdurbin) [inline lib/clandestined/rendezvous_hash.rb]: should we even bother assigning `score` as a variable here, or modify this to be

``` ruby
results << [hash_function.call("#{node}-#{key}"), node]
```

File: lib/clandestined/rendezvous_hash.rb [hunk 2]
```diff
@@ -29,18 +29,17 @@ def remove_node(node)
...
       end
-      winner
+      # Sort descending by score, use the node id converted to string as tiebraker
+      results.sort{|a,b| [b[0], b[1].to_s] <=> [a[0], a[1].to_s]}.take(num_nodes).map(&:last)
```
Comment (ewdurbin) [inline lib/clandestined/rendezvous_hash.rb:41]: this is _really_ dense and a great solution to the scoring/tiebreak. thanks for the comment :)

File: lib/clandestined/rendezvous_hash.rb [hunk 3]
```diff
@@ -29,18 +29,17 @@ def remove_node(node)
     end
 
     def find_node(key)
-      high_score = -1
-      winner = nil
+      find_nodes(key).first
+    end
+
+    def find_nodes(key, num_nodes=1)
```
Comment (ewdurbin) [inline lib/clandestined/rendezvous_hash.rb:35]: i'm not sure if the names `find_nodes` and is `num_nodes` are very descriptive for what's occurring here. while concise, it doesn't make it clear that you're getting back the "top `k` results in descending order by score".

do you have any alternative names for the method and argument?

Comment (swrobel) [inline lib/clandestined/rendezvous_hash.rb:35]: I don't - I was trying to stick with the convention, but I'm open to suggestions.

File: lib/clandestined/rendezvous_hash.rb (hunk 1 above)
Comment (swrobel) [inline lib/clandestined/rendezvous_hash.rb]: Good point! I made this change

**Comment by swrobel** [comment]
Any further thoughts on this?

**Pull request closed without merging**